RationaleMCP/0027/HindleyMilner.md

```diff
@@ -0,0 +1,336 @@
+They also provide the constraints that arise from different Modelica constructs.
+
+When it is specified that _`u1` must be equivalent to `u2`_, the corresponding constraint should be collected.
+
+In this section, the unit meta-expression of expression `e` is called `e.unit` and conditions of conditional constraints are highlighted with [].
+
+### Variables
+If `var` has a declared `unit`-attribute, `var.unit` must be equivalent to it.
+
+`var.unit` must be equivalent to the unit of the following attributes if present:
+ * `start`
+ * `min`
+ * `max`
+ * `nominal`
+
+(`var.unit` should be convertible to the `displayUnit`-attribute of `var` if it is present, but there is no corresponding constraints for the unit inference.)
+
+Consider the expression `e` of the form `var`, where `var` is a Modelica variable.
+* If `var` doesn't have a declared unit attribute, and has constant component variablity, `e.unit` is the `empty` unit.
+* Otherwise, `e.unit` is `var.unit`.
+
+### Literals
+`Real` literals have `empty` unit.
+
+### Binary operations
+Consider an expression `e` of the form `e1 op e2`.
+
+#### Additive operators
+* `e.unit` must be equivalent to `e1.unit`.
+* `e.unit` must be equivalent to `e2.unit`.
+
+Note that this requires introducing an intermediate variable to represent `e.unit`.
+This allows unit propagation when one of the operand has `empty` unit.
+
+#### Multiplication
+`e.unit` is `e1.unit * e2.unit`.
+
+#### Division
+`e.unit` is `e1.unit * (e2.unit ^ -1)`.
+
+#### Relational operators
+`e1.unit` must be equivalent to `e2.unit`
+
+#### Power operator
+* `e.unit` must be equivalent to `e1.unit ^ e2`.
+* If [`e2` wasn't evaluated or is a `Real` expression], `e1.unit` must be equivalent to `"1"`.
+
+Note that this requires introducing an intermediate variable to represent `e.unit`.
+Technically, `e2` is a Modelica expression rather than a rational number, which means that `e1.unit ^ e2` cannot be represented as a unit meta-expression.
+In practice, if the condition of the second constraint is verified, `e1.unit ^ e2` is `"1"` and if it isn't verified, then it can be represented as a meta-expression.
+
+(If the exponent was not evaluated or is a `Real`, the unit of the base should be `"1"`.)
+(It is up for discussion, when `e2` is `Real`, whether `e2.unit` should be equivalent to `"1"`. It could be handled in a similar way as transcendental functions. See below.)
+
+### Function calls
+Consider the expression `e` of the form `f(e1, e2, …, en)`.
+* If the output of the function has a declared unit, `e.unit` is the declared unit.
+* `e.unit` is `undefined` otherwise.
+
+If function input `i` has a declared unit, `ei.unit` must be equivalent to this unit.
+
+The natural generalization to functions with multiple outputs applies.
+
+### Transcendental functions
+As examples of transcendental functions, consider `sin`, `sign`, and `atan2`.
```

Comment:
The `sign` function is as far as I understand not a transcendental function, as they are defined as non-polynomial analytic functions, and I'm a bit unsure about `atan2`.

More importantly neither of those functions should use the rules for transcendental functions:
- The sign-function is just comparing a value to zero; the unit of the input doesn't matter.
- The atan2-function should ideally be defined as `atan2(y, x)=atan(y/x);` i.e. the two inputs should have the same unit - but it is perfectly normal to use atan2 for variables with unit.

Reply:
> * The sign-function is just comparing a value to zero; the unit of the input doesn't matter.

It does if it is a temperature.
 
> * The atan2-function should ideally be defined as `atan2(y, x)=atan(y/x);` i.e. the two inputs should have the same unit - but it is perfectly normal to use atan2 for variables with unit.

This is what is specified here.

Reply:
> > * The sign-function is just comparing a value to zero; the unit of the input doesn't matter.
> 
> It does if it is a temperature.

No.

Or with more words: it's more complicated - the unit sort of matters if it is an absolute temperature, but not if it is a temperature difference (and the unit system doesn't separate the two). And in the SI-system, which we base the unit handling on, absolute temperatures are measures in Kelvin and using the sign for an absolute temperature in Kelvin is kind of meaningless.

Similarly you should normally use sign for voltages, not potentials; and for relative positions (or velocities) - not absolute positions. The unit system doesn't separate them, and there are lots of models using sign for variables with units (fluid, mechanical, etc) - and they make sense.

Reply:
Fair enough for `sign`, I agree that we would need to ensures that the values is a difference.

Reply:
I just had another look at sign, the specification doesn't enforce any unit on its argument, it is just specifying that sign(-1) should behave as a literal.